### PR TITLE
Create env file and modify the Make file

### PR DESCRIPTION
**Pull request by tuananh1508**
This PR is to move all the hardcoded elements in this repo into variables.

The goal is to make it easier to deploy a Unee-T installation and avoid extensively suctomizing this repo each time we need to deploy a new Unee-T installation.

### CODE REVIEW

File: aws-env.dev [hunk 1]
```diff
@@ -0,0 +1,8 @@
+##this file store all evronment variables.
+PROJECT = lambda2sqs
+AWS_REGION = ap-southeast-1
+DEPLOY_S3_PREFIX = lambda2sqs
+AWS_PROFILE = uneet-dev
+DefaultSecurityGroup = sg-d8dde8a6
```
Comment (franck-boullier) [inline aws-env.dev]: @tuananh1508  can we use a consistent convention for the variable?
change `DefaultSecurityGroup` to `DEFAULT_SECURITY_GROUP`

File: aws-env.dev [hunk 2]
```diff
@@ -0,0 +1,8 @@
+##this file store all evronment variables.
```
Comment (franck-boullier) [inline aws-env.dev]: Replace with
`## This file store all environment variables for the DEV environment

**Review by franck-boullier — CHANGES_REQUESTED** [review]

**Comment by franck-boullier** [comment]
@kaihendry can you have a look at these changes and give us your input please?

File: aws-env.dev (hunk 1 above)
Comment (tuananh1508) [reply aws-env.dev]: Updated..

File: aws-env.dev [hunk 3]
```diff
@@ -0,0 +1,8 @@
+## This file store all environment variables for the DEV environment
+PROJECT = lambda2sqs
+AWS_REGION = ap-southeast-1
+DEPLOY_S3_PREFIX = lambda2sqs
+AWS_PROFILE = uneet-dev
```
Comment (kaihendry) [inline aws-env.dev]: Shouldn't that be ins-dev?

File: aws-env.dev [hunk 4]
```diff
@@ -0,0 +1,8 @@
...
+DEFAULT_SECURITY_GROUP = sg-d8dde8a6
+##PRIVATE_SUBNETS = Not Created yet.
```
Comment (kaihendry) [inline aws-env.dev]: without the subnets, this patch will break a potentially working setup, isn't it best just to add them?

**Comment by kaihendry** [comment]
Why is there a need to split out a env file? Why not have everything in one file?

**Comment by franck-boullier** [comment]
> Why is there a need to split out an env file? Why not have everything in one file?

We want to have consistency across repos.
We also want to standardize as many things as possible.

Eventually, we want the variables that are environment dependant to have the same name and values in this repo regardless of the environment that is being deployed (DEV, PROD or DEMO).

We also want to leverage the AWS parameter store as much as we can (like we do in the frontend and bugzilla-customization repos).

The end goal is for a given environment (DEV, PROD, DEMO) to only have to update 2 things:
- The aws-env.xxx file for this environment in the relevant repo (and leave all the other files untouched in the repo)
- The AWS parameter store for the environment.

This means that in the end, we will have to do more changes:

# New parameters in the AWS parameter store

We will create several new variables in the parameter store like
DEFAULT_SECURITY_GROUP
PRIVATE_SUBNET_1
PRIVATE_SUBNET_2
PRIVATE_SUBNET_3
INSTALLATION
DOMAIN

# changes to the file `aws-env.xxx`

(something similar to what we have in the frontend and bugzilla customization repo)

## New procedure:
In the file `aws-env.xxx` we'll create a procedure
```
getparam () {
	aws --profile ins-dev ssm get-parameters --names "$1" --with-decryption --query Parameters[0].Value --output text
}
```
## New variables: 
We will add 2 variables like
export INSTALLATION=$(getparam INSTALLATION)
export STAGE=dev

## Standardize other variables

AWS_PROFILE = $INSTALLATION-$STAGE
PRIVATE_SUBNETS = $PRIVATE_SUBNET_1,$PRIVATE_SUBNET_2,$PRIVATE_SUBNET_3

File: aws-env.dev (hunk 3 above)
Comment (franck-boullier) [reply aws-env.dev]: @kai is right, AWS profile is incorrect there.
My recommendation would be to define this variable as documented in my earlier comment: https://github.com/Unee-T-INS/lambda2sqs/pull/1#issuecomment-570032991

Comment (franck-boullier) [inline aws-env.dev]: @tuananh1508 if I'm not mistaken you should remove the spaces 
`PROJECT = lambda2sqs` --> `PROJECT=lambda2sqs`
`AWS_REGION = ap-southeast-1` --> `AWS_REGION=ap-southeast-1`

etc...

File: aws-env.dev (hunk 4 above)
Comment (franck-boullier) [inline aws-env.dev]: @tuananh1508 if I'm not mistaken you should remove the spaces 
`PROJECT = lambda2sqs` --> `PROJECT=lambda2sqs`
`AWS_REGION = ap-southeast-1` --> `AWS_REGION=ap-southeast-1`

etc...

**Comment by franck-boullier** [comment]
@tuananh1508, my take on the file `aws-env.dev`

```
## This file stores all environment variables for the DEV environment for this repo

# First, we create a procedure to retrieve the variables from the AWS parameter store
# Make sure to 
#   - configure your AWS CLI accordingly
#   - update the profile you need to use to access these variables if needed

    getparam () {
        aws --profile ins-dev ssm get-parameters --names "$1" --with-decryption --query Parameters[0].Value --output text
    }

# We prepare the hardcoded variables that we will need to deploy what's needed

    PROJECT=lambda2sqs
    DEPLOY_S3_PREFIX=lambda2sqs
    STAGE=dev

# Variables that are maintained in the AWS parameter store for the environment:

    AWS_REGION=$(getparam DEFAULT_REGION)
    DEFAULT_SECURITY_GROUP=$(getparam DEFAULT_SECURITY_GROUP)
    PRIVATE_SUBNET_1=$(getparam PRIVATE_SUBNET_1)
    PRIVATE_SUBNET_2=$(getparam PRIVATE_SUBNET_2)
    PRIVATE_SUBNET_3=$(getparam PRIVATE_SUBNET_3)

# Variables that are built from other variables:

    AWS_PROFILE=$(getparam INSTALLATION_ID)-$STAGE
    PRIVATE_SUBNETS=$PRIVATE_SUBNET_1,$PRIVATE_SUBNET_2,$PRIVATE_SUBNET_3
```

**Comment by franck-boullier** [comment]
@tuananh1508 and @kaihendry There are several other files that use hard coded parameters that will have to be modified or ideally moved into variables:

I have identified the following files with hardcoded variables:
-  https://github.com/Unee-T-INS/lambda2sqs/blob/master/.travis.yml 
- https://github.com/Unee-T-INS/lambda2sqs/blob/master/loglambdas.sh
- https://github.com/Unee-T-INS/lambda2sqs/blob/master/template.yaml

We probably need to edit thoses in order for the service to work as expected in the Unee-T INS installation.

**Comment by kaihendry** [comment]
My concern here is that the changes you might make may break CI https://travis-ci.org/unee-t/ and delay the goal of setting up a new environment.

aws-env.* is hardly a standard worth settling on since it's utility is diminished if you are going to store **every environment variable in the Parameter Store** aka ssm. https://aws.amazon.com/about-aws/whats-new/2018/11/aws-launches-secrets-support-for-amazon-elastic-container-servic/

You might as well just pull the secrets directly out in:
* https://github.com/unee-t/bugzilla-customisation/blob/master/AWS-docker-compose.yml
* https://github.com/unee-t/frontend/blob/master/AWS-docker-compose.yml

These changes might seem trivial, but one mistake and things can quite difficult to debug, especially if the value is in a parameter store.

**Review by franck-boullier — CHANGES_REQUESTED** [review]
Until we have a way to use variables instead of hard coded values, we need to replace
`uneet-dev`
with 
`ins-dev`

DONE in https://github.com/Unee-T-INS/lambda2sqs/pull/1/commits/e65ada06cb822920e6a1246f9adc8b330ac0c939

File: aws-env.dev [hunk 5]
```diff
@@ -0,0 +1,40 @@
...
+
+# First, we create a procedure to retrieve the variables from the AWS parameter store
+# Make sure to 
+#   - configure your AWS CLI accordingly
+#   - update the profile you need to use to access these variables if needed
+
+    getparam () {
+        #When we run this command, we get the issue: no matches found: Parameters[0].Value.
+        #aws --profile ins-dev ssm get-parameters --names "$1" --with-decryption --query Parameters[0].Value --output text
+        #So I changed the command like below, and It works!
```
Comment (franck-boullier) [inline aws-env.dev:12]: @kaihendry any idea why we have an error when we run this?

**Comment by franck-boullier** [comment]
@kaihendry can you have a look at the changes we've made with @tuananh1508 and let us know if you see any issues with that.

FYI, we have configured Travis-ci with the updated variable names and alse created a dedicated IAM user for travis to run the travis side of things.

File: .travis.yml [hunk 6]
```diff
@@ -11,6 +11,7 @@ install:
 deploy:
   # dev
   - provider: script
+  #TODO Explore if we can replace installation ID stage uneet-dev with variable 
     script: aws configure set profile.uneet-dev.aws_access_key_id $AWS_812644853088_ID &&
         aws configure set profile.uneet-dev.aws_secret_access_key $AWS_812644853088_SECRET &&
         aws configure set profile.uneet-dev.region ap-southeast-1 &&
```
Comment (franck-boullier) [inline .travis.yml]: @tuananh1508, we need to replace the Travis variable with the variables we have set in Travis Ci for this repo:
- change `$AWS_812644853088_ID` to `$AWS_ACCOUNT_USER_ID_DEV`
- change `$AWS_812644853088_SECRET` to `$AWS_ACCOUNT_SECRET_DEV`
- change `ap-southeast-1` to `$AWS_DEFAULT_REGION`

Comment (franck-boullier) [reply .travis.yml]: Done in https://github.com/Unee-T-INS/lambda2sqs/pull/1/commits/e65ada06cb822920e6a1246f9adc8b330ac0c939

**Pull request merged**